nf-core DSL2 compatibilty

spficklin opened the pull request:
This PR updates GEMmaker to compatibility with nf-core DSL2 framework.  The major changes are

1.  There is a new `modules/local` directory where all of the processes have been moved into a  separate files.  This isn't fully nf-core module worthy but our modules don't quite follow nf-core guidelines for sharability so this should be fine for now.  
2. The `main.nf` is just a short script that calls the GEMmaker workflow.
3. The GEMmaker workflow is now found here `workflows/GEMmaker.nf` 
4. The intialization steps for the GEMmaker workflow are found here `lib/WorkflowGEMmaker.groovy`
5. There are a lot of new files that the nf-core DSL 2.0 modules have and I've added those in and removed file that are no longer need.

Some things that are still off that we'll want to fix
1. For nf-core DSL2 each process is expected to have it's own Docker image.  Interestingly, this is the way we did it with GEMmaker initially and then switched to the single Docker image when we moved to nf-core.   For now, though, I've set each process to use a GEMmaker-2.0.0 docker image.  The `environment.yml` file is no longer needed since we are expected to use public existing Docker images for tools.  This might not quite work for GEMmaker because all of our tools have some "#TRACE" output in the stdout so we can't just use the public Docker images for reach tool. 
2.  The `retrieve_sra_metadata` never seems to get cached or for some reason Nextflow thinks something has changed and re-runs it.  This will always cause SRAs to be redownloaded every time.  This needs fixing before we make a release.  I can work on this bug once this PR gets merged.  I didn't want to make the PR much bigger.
3. If all of the samples complete but GEMmaker fails during the `create_gem` step it will never create the GEM when it resumes.  We need to fix this, but that's for a different PR.

bentsherman commented on modules/local/clean_work_dirs.nf:
```diff
@@ -0,0 +1,20 @@
+
+    script:
+    """
+    clean_work_dirs.sh "${directory}"
```
Since each process is in its own file now, would it make sense to move the bin scripts back into the processes? I think that would be better at this point.

spficklin replied:
Yes, I agree.  The modules should be self-contained and not rely on scripts in the bin directory.  I'll try and make this change.

spficklin replied:
I just pushed a commit that removed the .sh files and merged in that code to the module files.  I tested all three aligners and they worked.

bentsherman commented on modules/local/star.nf:
```diff
@@ -0,0 +1,37 @@
+params.options = [:]
+
+/**
+ * Performs hisat2 alignment of fastq files to a genome reference
```
Should say star instead of hisat2?

spficklin replied:
good catch!  I'll fix it.

bentsherman commented:
@spficklin For (1) I would use standard Docker images for processes if they already exist, and use the gemmaker image otherwise. The trace directives are all very simple, they should work either way. I would caution against over-optimizing for nf-core if it makes GEMmaker harder to improve and maintain.

As for (2) and (3), I think you should go ahead and investigate these issues in this PR. These changes are not urgent, so we can sit with it for a little while to make sure everything is stable. If the fixes for these problems turn out to be simple, then including them here is fine, but even if they are not, it might be worthwhile to wrestle with those changes while we're already re-organizing everything. I just don't see any reason to introduce two new bugs when this PR is mostly just housekeeping. Unless there are other big changes you want to make downstream that are enabled by this refactoring.

spficklin commented:
Okay, I fixed both bugs from (2) and (3).  The problem was that the processes that looked at the `work/GEMmaker` folder were using a file path to that directory. Nextflow is smart in that when it sees anything that looks like a path in a variable  in the `script` section, it will check its stats and if it sees the file/directory has changed then it will re-run that step.  Because the `work/GEMmaker` folder is always changing it meant those processes that look there don't ever use their cache but always get re-run. I had to trick Nextflow by passing in variables that didn't quite look like a path and then refactoring them.  You'll see what I did in the code.  

Also, another change I made was moving all sample files from both the `work/GEMmaker/process` and `work/GEMmaker/done` folders back into the `work/GEMmaker/stage` folder on a resume.   This doesn't cause any sample to get re-executed because they are already cached, but it does allow the samples to flow through the channels and Nextflow will print out all of the cached processes which fixes problem (3) and makes it nicer for the end-user because it's clear what's been done and what hasn't been done. 

Here's an example of the Hisat2 pipeline with caching of the retrieve and download working!  

![image](https://user-images.githubusercontent.com/1719352/142720607-39c5b161-0c24-4957-8832-ad1b55d99029.png)

I also changed the `container` for all of the processes with tools  that had their own compatible Docker  to use the public Docker image instead of the gemmaker-2.0.0 docker image.  There are still several processes that need the gemmaker image but it's a bit lighter now as it doesn't need so many tools in it.  

I have tested all three alignment tools and they are all working. 

There is still one issue, but it's not major. The `get_software_versions` process never runs. I think it's because the processes that use actual tools need to be more formally defined following the nf-core module standards. I think since it's working we can add a new issue for that and call this good....

bentsherman commented:
@spficklin I reviewed your updates and everything looks good to me. I'm glad you were able to fix the issues (2) and (3). The "path:" thing is a clever hack, I'm sure we'll be able to find a cleaner way to do it in the future.

I see what you meant now about the trace directives. We can't use the standard nf-core/modules because our versions have the trace directives in them. That is unfortunate. But I see the trace directives in a similar light as GEMmaker's cleanup code -- hopefully both of these features will eventually be incorporated into Nextflow itself.

Ready to merge when you are.

spficklin commented:
Okay! I just ran a test of the 475 rice dataset on Kamiak and it's having a big of issues on the fastq_merge step.  Moving the bash scripts back into the process has made it so that if any of the commands return a non-zero exit status nextflow stops.  If there is non-paired data then one of the `ls` commands returns a 1 exit status. I thought the fix I did worked to get around this. It does on my Ubuntu machine but not on Kamiak.  So, let me try and fix that and then I'll merge.  Thanks for the quick reviews.

spficklin commented:
This branch has processed over 4K samples on the 26K Arabidopsis dataset on Kamiak without issues. So, I'm merging.
